src/appleseed/renderer/modeling/bsdf/hairbsdf.cpp

```diff
@@ -0,0 +1,840 @@
+        {
+            float melanin = -std::log(std::max(1.f - melanin_conc, 0.0001f));
+            float ce = melanin * (1.f - melanin_ratio);
+            float cp = melanin * melanin_ratio;
+            Vector3f eumelanin_sigma_a = Vector3f(0.506f, 0.841f, 1.653f);
+            Vector3f pheomelanin_sigma_a = Vector3f(0.343f, 0.733f, 1.924f);
+            Vector3f sigma_a = (ce * eumelanin_sigma_a + cp * pheomelanin_sigma_a);
+            Spectrum ret(0.f);
+            ret.set(Color3f(sigma_a[0], sigma_a[1], sigma_a[2]), g_std_lighting_conditions, Spectrum::Intent::Reflectance);
+            return ret;
+        }
+
+        //
+        // Computes absorption coefficient based on azimuthal roughness and surface reflectance.
+        //
+
+        Spectrum sigma_a_from_rgb(const Spectrum& reflectance, float beta_n) const
+        {
+            Spectrum sigma_a(0.f);
+            for (int i = 0; i < reflectance.size(); i++)
+            {
+                float s = std::log(reflectance[i]) /
+                          (5.969f - 0.215f * beta_n + 2.532f * beta_n * beta_n -
+                           10.73f * pow_int(beta_n, 3) + 5.574f * pow_int(beta_n, 4) + 0.245f * pow_int(beta_n, 5));
+                sigma_a[i] = s * s;
+            }
+            return sigma_a;
+        }
+
+
+    private:
+        typedef HairBSDFInputValues InputValues;
+    };
+
+    typedef BSDFWrapper<HairBSDFImpl> HairBSDF;
+}
+
+
+//
+// HairBSDFFactory class implementation.
+//
+
+void HairBSDFFactory::release()
+{
+    delete this;
+}
+
+const char* HairBSDFFactory::get_model() const
+{
+    return Model;
+}
+
+Dictionary HairBSDFFactory::get_model_metadata() const
+{
+    return
+        Dictionary()
+            .insert("name", Model)
+            .insert("label", "Hair BSDF");
+}
+
+DictionaryArray HairBSDFFactory::get_input_metadata() const
+{
+    DictionaryArray metadata;
+
+    metadata.push_back(
+        Dictionary()
+            .insert("name", "reflectance")
+            .insert("label", "Reflectance")
+            .insert("type", "colormap")
+            .insert("entity_types",
+                    Dictionary()
+                        .insert("color", "Colors")
+                        .insert("texture_instance", "Textures"))
+            .insert("use", "required")
+            .insert("default", "0.00"));
+
+
+    metadata.push_back(
+        Dictionary()
+            .insert("name", "melanin")
+            .insert("label", "Melanin concentration")
+            .insert("type", "numeric")
+            .insert("min",
+                    Dictionary()
+                        .insert("value", "0.0")
+                        .insert("type", "hard"))
+            .insert("max",
+                    Dictionary()
+                        .insert("value", "1.0")
+                        .insert("type", "hard"))
+            .insert("use", "optional")
+            .insert("default", "0.0"));
```

Comment:
Defaults should match those used when declaring inputs. This duplication sucks but until we fix it, we have to keep things consistent.